### ISSUE TITLE
训练loss从30轮开始变为nan

### ISSUE THREAD

**Issue by orange-111222** [bowang-lab/U-Mamba#37]
2024-05-06 06:49:37.241385: Epoch 29 
2024-05-06 06:49:37.241527: Current learning rate: 0.00974 
2024-05-06 06:52:36.636003: train_loss -0.5247 
2024-05-06 06:52:36.636215: val_loss -0.6027 
2024-05-06 06:52:36.636309: Pseudo dice [0.5292, 0.0, 0.4787, nan, 0.0, 0.0] 
2024-05-06 06:52:36.636381: Epoch time: 179.4 s 
2024-05-06 06:52:37.962367:  
2024-05-06 06:52:37.962497: Epoch 30 
2024-05-06 06:52:37.962595: Current learning rate: 0.00973 
2024-05-06 06:55:37.388743: train_loss -0.4569 
2024-05-06 06:55:37.388954: val_loss -0.5265 
2024-05-06 06:55:37.389062: Pseudo dice [0.4189, 0.0, 0.3591, nan, 0.0, 0.0] 
2024-05-06 06:55:37.389136: Epoch time: 179.43 s 
2024-05-06 06:55:38.751563:  
2024-05-06 06:55:38.751762: Epoch 31 
2024-05-06 06:55:38.751870: Current learning rate: 0.00972 
2024-05-06 06:58:34.520405: train_loss nan 
2024-05-06 06:58:34.520603: val_loss nan 
2024-05-06 06:58:34.520698: Pseudo dice [0.0199, 0.0, 0.0022, nan, 0.0, 0.0] 
2024-05-06 06:58:34.520773: Epoch time: 175.77 s 
2024-05-06 06:58:35.810734:  
2024-05-06 06:58:35.810878: Epoch 32 
2024-05-06 06:58:35.810974: Current learning rate: 0.00971 
2024-05-06 07:01:28.559406: train_loss nan 
2024-05-06 07:01:28.559933: val_loss nan 
2024-05-06 07:01:28.560041: Pseudo dice [0.0, 0.0, 0.0, nan, 0.0, 0.0] 
2024-05-06 07:01:28.560115: Epoch time: 172.75 s 
2024-05-06 07:01:29.910507:  
2024-05-06 07:01:29.910718: Epoch 33 
2024-05-06 07:01:29.910861: Current learning rate: 0.0097 
2024-05-06 07:04:21.132929: train_loss nan 
2024-05-06 07:04:21.133433: val_loss nan 
2024-05-06 07:04:21.133681: Pseudo dice [0.0, 0.0, 0.0, nan, 0.0, 0.0] 
2024-05-06 07:04:21.133753: Epoch time: 171.22 s 
2024-05-06 07:04:22.612751:  
2024-05-06 07:04:22.612866: Epoch 34 
2024-05-06 07:04:22.612957: Current learning rate: 0.00969 
2024-05-06 07:07:13.875252: train_loss nan 
2024-05-06 07:07:13.875460: val_loss nan 
2024-05-06 07:07:13.875552: Pseudo dice [0.0, 0.0, 0.0, nan, 0.0, 0.0] 
2024-05-06 07:07:13.875623: Epoch time: 171.26 s 

![progress](https://github.com/bowang-lab/U-Mamba/assets/169115909/1f6d8c2a-dc4c-4cf7-af68-4ad38c620307)

**Comment by orange-111222** [comment]
我现在把lr降低到原来的1/10，训练变正常了

**Comment by orange-111222** [comment]
可能是因为我自己的数据集数量太小的原因吧

**Comment by Ayao29** [comment]
> 我现在把lr降低到原来的1/10，训练变正常了

笑死，我训练自己的数据集也遇到了这个问题了，也是这个方法就解决了

**Comment by Itsanewday** [comment]
> > 我现在把lr降低到原来的1/10，训练变正常了
> 
> 笑死，我训练自己的数据集也遇到了这个问题了，也是这个方法就解决了

最终训练完的模型效果咋样？与nnUNet中的resenc相比有提升吗?

**Comment by orange-111222** [comment]
> > > 我现在把lr降低到原来的1/10，训练变正常了
> > 
> > 
> > 笑死，我训练自己的数据集也遇到了这个问题了，也是这个方法就解决了
> 
> 最终训练完的模型效果咋样？与nnUNet中的resenc相比有提升吗?

训练属实慢，500轮的best不如nnUnet

**Comment by Itsanewday** [comment]
> > > > 我现在把lr降低到原来的1/10，训练变正常了
> > > 
> > > 
> > > 笑死，我训练自己的数据集也遇到了这个问题了，也是这个方法就解决了
> > 
> > 
> > 最终训练完的模型效果咋样？与nnUNet中的resenc相比有提升吗?
> 
> 训练属实慢，500轮的best不如nnUnet

谢谢，我之前用3090跑过几次ULS23数据集，都出现了问题，只有一次能正常训练，但是最终的dice也为0, 估计是超参数设置不合理。你那边是用什么显卡跑成功的？马骏博士用U-mamba拿到了ULS23的冠军，能训练U-mamaba的话还是可以尝试的。

**Comment by orange-111222** [comment]
> > > > > 我现在把lr降低到原来的1/10，训练变正常了
> > > > 
> > > > 
> > > > 笑死，我训练自己的数据集也遇到了这个问题了，也是这个方法就解决了
> > > 
> > > 
> > > 最终训练完的模型效果咋样？与nnUNet中的resenc相比有提升吗?
> > 
> > 
> > 训练属实慢，500轮的best不如nnUnet
> 
> 谢谢，我之前用3090跑过几次ULS23数据集，都出现了问题，只有一次能正常训练，但是最终的dice也为0, 估计是超参数设置不合理。你那边是用什么显卡跑成功的？马骏博士用U-mamba拿到了ULS23的冠军，能训练U-mamaba的话还是可以尝试的。

我这边用的是A100 80g，差不多500秒一轮

**Comment by Weixiang-Sun** [comment]
你好，为什么会有负的loss

**Comment by orange-111222** [comment]
我用的是nnUnet里面的dice+ce，他里面的损失返回值我记得是加了负号，然后就是负的loss

**Comment by OCEANOUXIN** [comment]
我也遇到这个问题了，是要降学习率吗

**Comment by roxyyyy7** [comment]
请问为什么loss是负的，一直找不到原因

**Comment by Younger330** [comment]
你好 我也遇到了类似的问题 我训练的预测值一直都是负数

**Comment by roxyyyy7** [comment]
> 你好 我也遇到了类似的问题 我训练的预测值一直都是负数

查一下nnunet的损失公式 带负号

**Comment by Younger330** [comment]
> > 你好 我也遇到了类似的问题 我训练的预测值一直都是负数
> 
> 查一下nnunet的损失公式 带负号

您好 我试了完全使用其损失公式DC_and_CE_loss，结果依旧是负数，请问您有什么建议吗[https://github.com/bowang-lab/U-Mamba/issues/63](my issue)

**Comment by roxyyyy7** [comment]
> > > 你好 我也遇到了类似的问题 我训练的预测值一直都是负数
> > 
> > 
> > 查一下nnunet的损失公式 带负号
> 
> 您好 我试了完全使用其损失公式DC_and_CE_loss，结果依旧是负数，请问您有什么建议吗[https://github.com/[/issues/63](https://github.com/bowang-lab/U-Mamba/issues/63)](my issue)

这我就不清楚了，DC_and_CE_loss的结果应该是正数

**Comment by wxmadm** [comment]
> > > 你好 我也遇到了类似的问题 我训练的预测值一直都是负数
> > 
> > 
> > 查一下nnunet的损失公式 带负号
> 
> 您好 我试了完全使用其损失公式DC_and_CE_loss，结果依旧是负数，请问您有什么建议吗[https://github.com/[/issues/63](https://github.com/bowang-lab/U-Mamba/issues/63)](my issue)


请问你搞明白为什么是负的了吗，我现在的损失也是从刚开始的正值变成负值了

**Comment by Younger330** [comment]
> > > > 你好 我也遇到了类似的问题 我训练的预测值一直都是负数
> > > 
> > > 
> > > 查一下nnunet的损失公式 带负号
> > 
> > 
> > 您好 我试了完全使用其损失公式DC_and_CE_loss，结果依旧是负数，请问您有什么建议吗[https://github.com/[/issues/63](https://github.com/bowang-lab/U-Mamba/issues/63)](my issue)
> 
> 请问你搞明白为什么是负的了吗，我现在的损失也是从刚开始的正值变成负值了

加个负号就行 或者你忘记了做softmax 但我遇到的不是这种问题

**Comment by DrunkerL** [comment]
请问一下，默认训练的轮数是多少啊？

**Comment by wxmadm** [comment]
1000轮

**Comment by liuwenjun-test** [comment]
using pin_memory on device 0
2025-04-18 19:06:52.807117: train_loss nan
2025-04-18 19:06:52.813793: val_loss nan
2025-04-18 19:06:52.814002: Pseudo dice [0.0, 0.0, 0.0]
2025-04-18 19:06:52.814368: Epoch time: 597.65 s
2025-04-18 19:06:52.814471: Yayy! New best EMA pseudo Dice: 0.0

我的是原来一直可以正常运行，但是由于我不断地运行，停止，运行，停止，然后就出现了NAN，但是重启服务器就会可以再次正常运行，这是什么原因？也不可能总是重启服务器吧

**Comment by cjf128** [comment]
请问一下，效果上跟Bot对比如何？我用的NoAMP版本，显示显存不足，我现在想直接用普通的Enc，降低学习率来训练，不知道跟没降学习率的Bot对比怎么样
